### ISSUE TITLE
Support scanning multiple ranges

### ISSUE THREAD

**Issue by breeswish** [tikv/tikv#4239]
## Feature Request

Scanning multiple ranges (in Coprocessor) is not efficient currently because for each range we need to rebuild the rocksdb iterator. This results in a strange case that scanning from a to z can be faster than scanning from a to c, from f to g, from x to z.

We can support scanning multiple (ordered) ranges in the mvcc scanner, by using the same rocksdb iterator. Each time for a new range, we can next() for several times and then seek() for it, just like what we did in current mvcc process.

**Comment by BusyJay** [comment]
If there are a lot of tombstone keys between ranges, reusing iterator may perform worse.

**Comment by TennyZhuang** [comment]
How about creating a iterator pool in memory and always find the nearest iterator?

**Comment by breeswish** [comment]
@zhangjinpeng1987 PTAL

**Comment by zhangjinpeng1987** [comment]
@breeswish Can you provide more details for what kind of SQL can send queries that scan multiple ranges in one Coprocessor query.

**Comment by breeswish** [comment]
@zhangjinpeng1987 There are many, for example, `select * from table where id not in (...)`.

**Comment by zhangjinpeng1987** [comment]
I think it worth to reuse iterator in a coprocessor request.

**Comment by breeswish** [comment]
@zhangjinpeng1987 Yes I agree. It's this issue's purpose. Also @ilovesoup has some demands of scanning multiple ranges.
